storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbDatabase.java

```diff
@@ -266,12 +276,53 @@ public void storeFinalizedBlocks(final Collection<SignedBeaconBlock> blocks) {
       expectedRoot = block.getParentRoot();
     }
 
+    // Check block signatures are valid for blocks except the genesis block
+    Collection<SignedBeaconBlock> nonGenesisBlocks =
+        blocks.stream()
+            .filter(block -> !block.getSlot().equals(UInt64.ZERO))
+            .collect(Collectors.toList());
+    checkArgument(
+        batchVerifyHistoricalBlockSignatures(nonGenesisBlocks), "Block signatures are invalid");
+
     try (final FinalizedUpdater updater = finalizedDao.finalizedUpdater()) {
       sorted.forEach(updater::addFinalizedBlock);
       updater.commit();
     }
   }
 
+  boolean batchVerifyHistoricalBlockSignatures(final Collection<SignedBeaconBlock> blocks) {
+    BeaconState finalizedState =
+        this.hotDao
+            .getLatestFinalizedState()
+            .orElseThrow(
+                () ->
+                    new IllegalStateException(
+                        "Finalized state is required to sync finalized historical blocks"));
+    List<BLSSignature> signatures = new ArrayList<>();
+    List<Bytes> signingRoots = new ArrayList<>();
+    List<List<BLSPublicKey>> proposerPublicKeys = new ArrayList<>();
+
+    // TODO: This domain is dependent on the fork version. Thus when we support forks, we're going
+    // to have to change the way we retrieve the domain here.
```

Comment:
We should create a ticket for it and probably work out how to deal with this pretty soon since the first fork is coming and it will definitely affect this.

Reply:
Done: https://github.com/ConsenSys/teku/issues/3430